modules/openy_gc_auth/src/GCUserAuthorizer.php

```diff
@@ -27,58 +33,137 @@ class GCUserAuthorizer {
+   *
+   * @return bool
+   *   True if the user successfully logged in.
    */
-  public function authorizeUser($name, $email, array $extra_data = []) {
+  public function authorizeUser($name, $mail, array $extra_data = []) {
 
-    if (empty($name) || empty($email)) {
-      return;
+    if (empty($name) || empty($mail)) {
+      return FALSE;
     }
 
-    // Create drupal user if it doesn't exist and login it.
-    $account = user_load_by_mail($email);
-    if (!$account) {
-      $user = $this->userStorage->create();
-      $user->setPassword(user_password());
-      $user->enforceIsNew();
-      $user->setEmail($email);
-      $user->setUsername($name);
-      $user->addRole(self::VIRTUAL_Y_DEFAULT_ROLE);
-      $user->activate();
-      $result = $account = $user->save();
-      if ($result) {
-        $account = user_load_by_mail($email);
+    $account = NULL;
+    // Load drupal user by name and email from SalesForce.
+    $account_by_mail = user_load_by_mail($mail);
+    $account_by_name = user_load_by_name($name);
+
+    // If we can load at least one account by userdata(from Salesforce).
+    if ($account_by_mail || $account_by_name) {
+
+      // If we have two account loaded, we should check if it's one user or two
+      // different(f.e. case if user change email in service provider, but it's
+      // already used in Drupal).
+      if ($account_by_mail && $account_by_name) {
+        // If for email and name from service provider registered two different
+        // account. Should be resolved manually in Drupal.
+        if ($account_by_mail->id() !== $account_by_name->id()) {
+          $this->resolveAccountConflict($name, $mail);
+        }
+        // If both account point to same User.
+        else {
+          // Doesn't matter, what object to use, as they are the same.
+          $account = $account_by_mail;
+        }
       }
-    }
-    else {
-      // Activate user if it's not.
-      if (!$account->isActive()) {
-        $account->activate();
-        $account->setPassword(user_password());
+      // Loaded one account -> email or name was changed in Service provider.
+      else {
+        // Get existing(loaded) account, that will be logged in.
+        $account = $account_by_mail ? $account_by_mail : $account_by_name;
+        // Get changed in Service provider field, update and log changing of
+        // related field in Drupal.
+        // If $account_by_mail exist, this mean that by name we didn't load user
+        // and name was changed in Service provider.
+        $changed_field = $account_by_mail ? 'name' : 'mail';
+        $old_changed_field_value = $account->get($changed_field)->value;
+        // Set new value (from service provider).
+        $account->set($changed_field, $$changed_field);
         $account->save();
+
+        $this->logger->error('Service provider user credentials was changed. For User with id:%id changed field \'%field\' from \'%old\' to \'%new\' value',
```

Comment:
Why should it be an error?... I believe we should use lower level here, like message or notice